File: targets/java-spring/services/FileService.java
package com.baddemo.services;

import java.util.Set;

import com.baddemo.config.AppConfig;
import com.baddemo.utils.FileUtil;

/** File retrieval logic. */
public class FileService {

    private static final Set<String> ALLOWED = Set.of("readme.txt", "help.txt");

    /** Passes user-controlled name down to the filesystem helper (sink hop). */
    public String readUserFile(String name) throws Exception {
        return FileUtil.read(AppConfig.UPLOAD_DIR, name);
    }

    public String readWhitelisted(String name) throws Exception {
        if (!ALLOWED.contains(name)) {
            throw new IllegalArgumentException("file not allowed");
        }
        return FileUtil.read(AppConfig.UPLOAD_DIR, name);
    }
}

File: targets/php-laravel/app/Services/FileService.php
<?php

namespace App\Services;

class FileService
{
    private const ALLOWED_FILES = ['readme.txt', 'help.txt'];

    private string $uploadDir;

    public function __construct()
    {
        $config = require __DIR__ . '/../../config/app.php';
        $this->uploadDir = $config['upload_dir'];
    }

    public function readUserFile(string $name): string
    {
        $path = $this->uploadDir . '/' . $name;
        return file_get_contents($path);
    }

    public function readWhitelisted(string $name): string
    {
        if (!in_array($name, self::ALLOWED_FILES, true)) {
            throw new \InvalidArgumentException('file not allowed');
        }
        $path = $this->uploadDir . '/' . $name;
        return file_get_contents($path);
    }
}
